File: js/emit_denominators.js
function
emit_denominators(u, p)
{
	var n, q, v;

	n = count_denominators(p);

	v = {type:LINE, a:[], level:u.level};

	p = cdr(p);
	q = car(p);

	if (isrational(q)) {
		if (q.b != 1) {
			emit_roman_text(v, q.b.toFixed(0));
			n++;
		}
		p = cdr(p);
	}

	while (iscons(p)) {

		q = car(p);
		p = cdr(p);

		if (!isdenominator(q))
			continue;

		if (v.a.length > 0)
			emit_medium_space(v);

		if (isminusone(caddr(q))) {
			q = cadr(q);
			if (car(q) == symbol(ADD) && n == 1)
				emit_expr(v, q); // parens not needed
			else
				emit_factor(v, q);
		} else {
			emit_base(v, cadr(q));
			emit_numeric_exponent(v, caddr(q)); // sign is not emitted
		}
	}

	if (v.a.length == 1)
		v = v.a[0];
	else
		emit_update(v);

	u.den = v;
}

File: app/emit_denominators.c
#include "app.h"

void
emit_denominators(struct atom *p)
{
	int n, t;
	char *s;
	struct atom *q;

	t = tos;
	n = count_denominators(p);
	p = cdr(p);

	while (iscons(p)) {

		q = car(p);
		p = cdr(p);

		if (!isdenominator(q))
			continue;

		if (tos > t)
			emit_medium_space();

		if (isrational(q)) {
			s = mstr(q->u.q.b);
			emit_roman_string(s);
			continue;
		}

		if (isminusone(caddr(q))) {
			q = cadr(q);
			if (car(q) == symbol(ADD) && n == 1)
				emit_expr(q); // parens not needed
			else
				emit_factor(q);
		} else {
			emit_base(cadr(q));
			emit_numeric_exponent(caddr(q)); // sign is not emitted
		}
	}

	emit_update_list(t);
}
